probe/controls/controls.go

```diff
@@ -6,33 +6,138 @@ import (
 )
 
-var (
-	mtx      = sync.Mutex{}
-	handlers = map[string]xfer.ControlHandlerFunc{}
-)
+// HandlerRegistryBackend is an interface for storing control request
+// handlers.
+type HandlerRegistryBackend interface {
+	// Lock locks the backend, so the batch insertions or
+	// removals can be performed.
+	Lock()
+	// Unlock unlocks the registry.
+	Unlock()
+	// Register a new control handler under a given
+	// id. Implementations should not call Lock() or Unlock()
+	// here, it will be done by HandlerRegistry.
+	Register(control string, f xfer.ControlHandlerFunc)
+	// Rm deletes the handler for a given name. Implementations
+	// should not call Lock() or Unlock() here, it will be done by
+	// HandlerRegistry.
+	Rm(control string)
+	// Handler gets the handler for a control. Implementations
+	// should not call Lock() or Unlock() here, it will be done by
+	// HandlerRegistry.
+	Handler(control string) (xfer.ControlHandlerFunc, bool)
+}
+
+type defaultBackend struct {
+	handlers map[string]xfer.ControlHandlerFunc
+	mtx      sync.Mutex
+}
+
+// NewDefaultHandlerRegistryBackend creates a default backend for
+// handler registry.
+func NewDefaultHandlerRegistryBackend() HandlerRegistryBackend {
+	return &defaultBackend{
+		handlers: map[string]xfer.ControlHandlerFunc{},
+	}
+}
+
+// Lock locks the registry, so the batch insertions or
+// removals can be performed.
+func (b *defaultBackend) Lock() {
+	b.mtx.Lock()
+}
+
+// Unlock unlocks the registry.
+func (b *defaultBackend) Unlock() {
+	b.mtx.Unlock()
+}
+
+// Register a new control handler under a given id.
+func (b *defaultBackend) Register(control string, f xfer.ControlHandlerFunc) {
+	b.handlers[control] = f
+}
+
+// Rm deletes the handler for a given name.
+func (b *defaultBackend) Rm(control string) {
+	delete(b.handlers, control)
+}
+
+// Handler gets the handler for a control.
+func (b *defaultBackend) Handler(control string) (xfer.ControlHandlerFunc, bool) {
+	handler, ok := b.handlers[control]
+	return handler, ok
+}
+
+// HandlerRegistry uses backend for storing and retrieving control
+// requests handlers.
+type HandlerRegistry struct {
+	backend HandlerRegistryBackend
+}
+
+// NewDefaultHandlerRegistry creates a registry with a default
+// backend.
+func NewDefaultHandlerRegistry() *HandlerRegistry {
+	return NewHandlerRegistry(NewDefaultHandlerRegistryBackend())
+}
+
+// NewHandlerRegistry creates a registry with a custom backend.
+func NewHandlerRegistry(backend HandlerRegistryBackend) *HandlerRegistry {
+	return &HandlerRegistry{
+		backend: backend,
+	}
+}
+
+// Register registers a new control handler under a given id.
+func (r *HandlerRegistry) Register(control string, f xfer.ControlHandlerFunc) {
+	r.backend.Lock()
+	defer r.backend.Unlock()
+	r.RegisterLocked(control, f)
+}
+
+// Rm deletes the handler for a given name.
+func (r *HandlerRegistry) Rm(control string) {
+	r.backend.Lock()
+	defer r.backend.Unlock()
+	r.RmLocked(control)
+}
```

Comment:
registry backend was for storing, and retrieving handlers and registry was for actually calling the handler and having a consistent error reporting.